File: LeetCode/1697. Checking Existence of Edge Length Limited Paths/solution.py
# class Solution:
#     def distanceLimitedPathsExist(self, n: int, edgeList: List[List[int]], queries: List[List[int]]) -> List[bool]:
#         m = len(queries)
#         edgeList.append([-1,-1,float('inf')])
#         edgeList.sort(key=lambda x: x[-1])
#         idx = sorted([i for i in range(m)], key=lambda x: queries[x][-1])
#         res = [False]*m
#         parent = [i for i in range(n)]
#         def find(cur):
#             if parent[cur] != cur:
#                 parent[cur] = find(parent[cur])
#             return parent[cur]
#         def union(a,b):
#             ra, rb = find(a), find(b)
#             if ra != rb:
#                 parent[ra] = rb
#                 return False
#             return True
#         pos = 0
#         for u, v, e in edgeList:
#             while pos < m and queries[idx[pos]][-1] <= e:
#                 p, q, _ = queries[idx[pos]]
#                 if find(p) == find(q):
#                     res[idx[pos]] = True
#                 pos += 1

#             if u >= 0 and v >= 0:
#                 union(u,v)
#         return res 
                


class Solution:
    def distanceLimitedPathsExist(self, n: int, edgeList: List[List[int]], queries: List[List[int]]) -> List[bool]:
        m = len(queries)
        # edgeList.append([-1,-1,float('inf')])
        edgeList.sort(key=lambda x: x[-1])
        idx = sorted([i for i in range(m)], key=lambda x: queries[x][-1])
        res = [False]*m
        parent = [i for i in range(n)]
        def find(cur):
            if parent[cur] != cur:
                parent[cur] = find(parent[cur])
            return parent[cur]
        def union(a,b):
            ra, rb = find(a), find(b)
            if ra != rb:
                parent[ra] = rb
                return False
            return True
        pos = 0
        for i in range(m):
            p, q, limit = queries[idx[i]]
            while pos < len(edgeList) and edgeList[pos][-1] < limit:
                u, v, _ = edgeList[pos]
                union(u,v)
                pos += 1

            if find(p) == find(q):
                res[idx[i]] = True
                
        return res 
                

File: LeetCode/1697. Checking Existence of Edge Length Limited Paths/solution.cpp
class Solution {
public:
    vector<bool> distanceLimitedPathsExist(int n, vector<vector<int>>& edgeList, vector<vector<int>>& queries) {
        sort(edgeList.begin(), edgeList.end(), [&](vector<int> &a, vector<int> &b){
            return a.back() < b.back();
        });
        int m = queries.size();
        vector<int> idx, parent(n);
        vector<bool> ans(m);
        for (int i = 0; i < m; i++)
            idx.push_back(i);
        sort(idx.begin(), idx.end(), [&](int a, int b){
            return queries[a].back() < queries[b].back();
        });
        for (int i = 0; i < n; i++)
            parent[i] = i;

        function<int(int)> find = [&](int root) -> int {
            if (parent[root] != root)
                parent[root] = find(parent[root]);
            return parent[root];
        };

        function<void(int,int)> merge = [&](int a, int b) -> void {
            int ra = find(a), rb = find(b);
            parent[ra] = rb;
            return;
        };
        int j = 0;
        for (auto i : idx){
            for (; j < edgeList.size() && edgeList[j].back() < queries[i].back(); j++){
                merge(edgeList[j][0], edgeList[j][1]);
            }
            if (find(queries[i][0]) == find(queries[i][1]))
                ans[i] = true;
            else
                ans[i] = false;
        }
        return ans;
    }
};



class Solution {
public:
    vector<bool> distanceLimitedPathsExist(int n, vector<vector<int>>& edgeList, vector<vector<int>>& queries) {
        sort(edgeList.begin(), edgeList.end(), [&](vector<int> &a, vector<int> &b){
            return a.back() < b.back();
        });
        int m = queries.size();
        vector<int> idx, parent(n), sz(n,1);
        vector<bool> ans(m);
        for (int i = 0; i < m; i++)
            idx.push_back(i);
        sort(idx.begin(), idx.end(), [&](int a, int b){
            return queries[a].back() < queries[b].back();
        });
        for (int i = 0; i < n; i++)
            parent[i] = i;

        function<int(int)> find = [&](int root) -> int {
            if (parent[root] != root)
                parent[root] = find(parent[root]);
            return parent[root];
        };

        function<void(int,int)> merge = [&](int a, int b) -> void {
            int ra = find(a), rb = find(b);
            if (ra == rb)
                return;
            if (sz[ra] < sz[rb]){
                parent[ra] = rb;
                sz[rb] += sz[ra];
            }
            else{
                parent[rb] = ra;
                sz[ra] += sz[rb];
            }
            return;
        };
        int j = 0;
        for (auto i : idx){
            for (; j < edgeList.size() && edgeList[j].back() < queries[i].back(); j++){
                merge(edgeList[j][0], edgeList[j][1]);
            }
            if (find(queries[i][0]) == find(queries[i][1]))
                ans[i] = true;
            else
                ans[i] = false;
        }
        return ans;
    }
};
